File: hmh-backend/app/schemas/programme.py
"""Pydantic schemas for Programme Activities."""

from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel

from app.models.enums import ProgrammeActivityStatus, ProgrammeActivityType


class ProgrammeActivityRead(BaseModel):
    id: uuid.UUID
    activity_number: str
    project_id: uuid.UUID
    site_id: Optional[uuid.UUID] = None
    lot_id: Optional[uuid.UUID] = None
    stage_status_id: Optional[uuid.UUID] = None
    title: str
    description: Optional[str] = None
    activity_type: str
    planned_start_date: date
    planned_finish_date: date
    actual_start_date: Optional[date] = None
    actual_finish_date: Optional[date] = None
    baseline_start_date: Optional[date] = None
    baseline_finish_date: Optional[date] = None
    duration_days: Optional[int] = None
    progress_pct: int
    status: str
    predecessor_id: Optional[uuid.UUID] = None
    lag_days: int
    is_critical_path: bool
    is_milestone: bool
    responsible_team: Optional[str] = None
    notes: Optional[str] = None
    created_by: Optional[uuid.UUID] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ProgrammeActivityCreate(BaseModel):
    title: str
    description: Optional[str] = None
    activity_type: str = ProgrammeActivityType.CONSTRUCTION.value
    site_id: Optional[uuid.UUID] = None
    lot_id: Optional[uuid.UUID] = None
    stage_status_id: Optional[uuid.UUID] = None
    planned_start_date: date
    planned_finish_date: date
    predecessor_id: Optional[uuid.UUID] = None
    lag_days: int = 0
    is_critical_path: bool = False
    is_milestone: bool = False
    responsible_team: Optional[str] = None
    notes: Optional[str] = None


class ProgrammeActivityUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    activity_type: Optional[str] = None
    site_id: Optional[uuid.UUID] = None
    lot_id: Optional[uuid.UUID] = None
    stage_status_id: Optional[uuid.UUID] = None
    planned_start_date: Optional[date] = None
    planned_finish_date: Optional[date] = None
    actual_start_date: Optional[date] = None
    actual_finish_date: Optional[date] = None
    baseline_start_date: Optional[date] = None
    baseline_finish_date: Optional[date] = None
    progress_pct: Optional[int] = None
    status: Optional[str] = None
    predecessor_id: Optional[uuid.UUID] = None
    lag_days: Optional[int] = None
    is_critical_path: Optional[bool] = None
    is_milestone: Optional[bool] = None
    responsible_team: Optional[str] = None
    notes: Optional[str] = None


class SetBaselineRequest(BaseModel):
    """Freeze current planned dates as immutable baseline."""
    confirm: bool = False

File: hmh-frontend/src/api/programme.ts
import client from "./client";

export type ProgrammeActivityStatus =
  | "NOT_STARTED" | "IN_PROGRESS" | "DELAYED" | "BLOCKED" | "COMPLETED" | "VERIFIED" | "CANCELLED";

export type ProgrammeActivityType =
  | "CONSTRUCTION" | "PROCUREMENT" | "INSPECTION" | "ADMIN" | "MILESTONE";

export interface ProgrammeActivity {
  id: string;
  activity_number: string;
  project_id: string;
  site_id: string | null;
  lot_id: string | null;
  stage_status_id: string | null;
  title: string;
  description: string | null;
  activity_type: ProgrammeActivityType;
  planned_start_date: string;
  planned_finish_date: string;
  actual_start_date: string | null;
  actual_finish_date: string | null;
  baseline_start_date: string | null;
  baseline_finish_date: string | null;
  duration_days: number | null;
  progress_pct: number;
  status: ProgrammeActivityStatus;
  predecessor_id: string | null;
  lag_days: number;
  is_critical_path: boolean;
  is_milestone: boolean;
  responsible_team: string | null;
  notes: string | null;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface ProgrammeActivityCreate {
  title: string;
  activity_type?: ProgrammeActivityType;
  site_id?: string | null;
  lot_id?: string | null;
  stage_status_id?: string | null;
  planned_start_date: string;
  planned_finish_date: string;
  predecessor_id?: string | null;
  lag_days?: number;
  is_critical_path?: boolean;
  is_milestone?: boolean;
  responsible_team?: string | null;
  description?: string | null;
  notes?: string | null;
}

export const listActivities = (projectId: string, params?: { site_id?: string; status?: string }) =>
  client.get(`/api/v1/projects/${projectId}/programme`, { params }).then((r) => r.data.data as ProgrammeActivity[]);

export const getActivity = (activityId: string) =>
  client.get(`/api/v1/programme/${activityId}`).then((r) => r.data.data as ProgrammeActivity);

export const createActivity = (projectId: string, body: ProgrammeActivityCreate) =>
  client.post(`/api/v1/projects/${projectId}/programme`, body).then((r) => r.data.data as ProgrammeActivity);

export const updateActivity = (activityId: string, body: Partial<ProgrammeActivityCreate> & { progress_pct?: number; status?: string; actual_start_date?: string; actual_finish_date?: string }) =>
  client.patch(`/api/v1/programme/${activityId}`, body).then((r) => r.data.data as ProgrammeActivity);

export const deleteActivity = (activityId: string) =>
  client.delete(`/api/v1/programme/${activityId}`).then((r) => r.data);

export const setBaseline = (activityId: string) =>
  client.post(`/api/v1/programme/${activityId}/baseline`, { confirm: true }).then((r) => r.data.data as ProgrammeActivity);
